[ADDED] js_PullSubscribeAsync

levb opened the pull request:
A replacement for https://github.com/nats-io/nats.c/pull/773 that seems to have lost its proper base in GitHub after changing it.

levb commented:
@kozlovic Travis seems to have frozen, I'll check it later, passed previously.

kozlovic commented on examples/js-sub.c:
```diff
@@ -130,21 +169,47 @@ int main(int argc, char **argv)
+            jsOpts.PullSubscribeAsync.FetchSize = 17;
+            jsOpts.PullSubscribeAsync.KeepAhead = 7;
+            jsOpts.PullSubscribeAsync.CompleteHandler = _completeFetchCb;
+            jsFetchRequest lifetime = {
```
Apparently, on my Windows VM, this type of initialization causes warning "C4204: nonstandard extension used: non-constant aggregate initializer".
One way to solve would be:
```
jsFetchRequest lifetime;

jsFetchRequest_Init(&lifetime);
lifetime.Batch   = (int) total;
lifetime.NoWait = true;
lifetime.Expires = 1 * HOUR_NANO;
```

levb replied:
Interesting why are we not getting it in CI with a /W4 (same [c]makefile anyway). Different compiler versions? I'll change and try to stay away from the pattern.

kozlovic replied:
It may be my (very old) Windows VM. Talking about `/W4`, when I compile on Windows, I get a warning, for each file compiled, saying "overriding `/W3` with `/W4`.

But again, I would not be surprised that it is due to my old Windows VM. If there is no complaint in the Windows build we have on GA, maybe ignore my comment?

levb replied:
I changed it.

In general, I would love to turn on warnings == errors on all platforms, but there are a few warnings (including disabling /W4 in test.c and now on your VM) that would get in the way

kozlovic commented on examples/js-sub.c:
```diff
@@ -130,21 +169,47 @@ int main(int argc, char **argv)
+                .Batch = total,
+                .NoWait = true,
+                .Expires = 1 * HOUR_NANO,
+                // .Heartbeat = 1 * SECOND_NANO,
```
Remove?

kozlovic commented on src/dispatch.c:
```diff
@@ -82,3 +83,452 @@ natsStatus natsSub_enqueueUserMessage(natsSubscription *sub, natsMsg *msg)
     return NATS_OK;
 }
 
+static inline void _removeFromDispatchQueue(natsDispatcher *d, natsMsg *msg)
```
Don't know if this was done before stopping the "format on save" with the clang-formatter, but if you want/could, I would have the return type in one line and the the function name in the next, to be consistent with (most of) the rest of the library.

kozlovic commented on src/dispatch.c:
```diff
@@ -82,3 +83,452 @@ natsStatus natsSub_enqueueUserMessage(natsSubscription *sub, natsMsg *msg)
+    d->queue.head = msg->next;
+    if (d->queue.tail == msg)
+        d->queue.tail = NULL;
+    msg->next = NULL;
```
The name of the function is not really indicative that this message needs to be the head of the list. I understand that we have selected the head before calling this function, and so this is just updating the list, but the fact that this function is named "remove" and we pass the message, one could think that the function is going to remove a message anywhere in the queue.

kozlovic commented on src/dispatch.c:
```diff
@@ -82,3 +83,452 @@ natsStatus natsSub_enqueueUserMessage(natsSubscription *sub, natsMsg *msg)
+}
+
+// Returns fetch status
+static inline natsStatus _preProcessUserMessage(
```
Same comment for return type.

kozlovic commented on src/dispatch.c:
```diff
@@ -82,3 +83,452 @@ natsStatus natsSub_enqueueUserMessage(natsSubscription *sub, natsMsg *msg)
+
+// Returns fetch status
+static inline natsStatus _preProcessUserMessage(
+    natsSubscription *sub, jsSub *jsi, jsFetch *fetch, natsMsg *msg,
```
May want to indicate if the `sub` lock must be held on entry or not.

kozlovic commented on src/dispatch.c:
```diff
@@ -82,3 +83,452 @@ natsStatus natsSub_enqueueUserMessage(natsSubscription *sub, natsMsg *msg)
+
+    // Fetch-specific handling of synthetic and header-only messages
+    if (fetch != NULL)
+        fetchStatus = js_checkFetchedMsg(sub, msg, true, userMsg);
```
This function says that the lock must be held, but looks like the caller of the current function is not in some cases (in cats_dispatchThreadPool).

levb replied:
Thanks for the catch, restructured so that no locking required.

kozlovic commented on src/dispatch.c:
```diff
@@ -82,3 +83,452 @@ natsStatus natsSub_enqueueUserMessage(natsSubscription *sub, natsMsg *msg)
+        _removeFromDispatchQueue(d, msg);
+
+        // Get subscription reference from message and capture values we need
+        // while under lock.
```
Not the sub's lock as far as I can tell.

levb replied:
Right, but all these should be protected by the dispatcher lock. This goes back to https://github.com/nats-io/nats.c/blob/v3.8.2/src/nats.c#L1698. Do you find the comment ambiguous, or do you see an issue with the code?

(Separate thread for closed/drained flags noted)

kozlovic commented on src/dispatch.c:
```diff
@@ -82,3 +83,452 @@ natsStatus natsSub_enqueueUserMessage(natsSubscription *sub, natsMsg *msg)
+
+            nats_lockDispatcher(d);
+
+            if (!draining && !closed)
```
We have released/reacquire the lock, but these are evaluated before. The situation may have changed, but I guess the state will be discovered at the next iteration(s) possibly?
The question is: is the subscription protected in case it is closed while the dispatcher lock is released?

levb replied:
The sub timeout is a rare event, so 0/5 I should lock the sub and re-acquire these values?

levb replied:
Removed the cached variables since the checking is already under the lock.

kozlovic commented on src/dispatch.c:
```diff
@@ -82,3 +83,452 @@ natsStatus natsSub_enqueueUserMessage(natsSubscription *sub, natsMsg *msg)
+            natsMsg_Destroy(msg);
+            continue;
+        }
+        else if ((msg->sub == NULL) || (msg->subject == NULL) || (strcmp(msg->subject, "") == 0))
```
I guess we could have combined with the "if" above since the action is the same, destroy message and continue.

kozlovic commented on src/dispatch.c:
```diff
@@ -82,3 +83,452 @@ natsStatus natsSub_enqueueUserMessage(natsSubscription *sub, natsMsg *msg)
+}
+
+// Thread main function for a thread pool of dispatchers.
+void nats_dispatchThreadPool(void *arg)
```
return type formatting?

kozlovic commented on src/dispatch.c:
```diff
@@ -82,3 +83,452 @@ natsStatus natsSub_enqueueUserMessage(natsSubscription *sub, natsMsg *msg)
+}
+
+// Thread main function for a dedicated dispatcher.
+void nats_dispatchThreadDedicated(void *arg)
```
Return type formatting?

kozlovic commented on src/dispatch.c:
```diff
@@ -82,3 +83,452 @@ natsStatus natsSub_enqueueUserMessage(natsSubscription *sub, natsMsg *msg)
+    void *completeCBClosure = NULL;
+
+    // This just serves as a barrier for the creation of this thread.
+    natsConn_Lock(nc);
```
I wonder if instead of having those we should have the sub lock/unlock when getting the fields above, even though we say that they are immutable.

levb replied:
It's not even the locking, no need to refresh them every iteration of the loop; and since they are already set when the thread is started and immutable after, I grouped them before the loop. Change the comment?

kozlovic replied:
Actually, I added that back [here](https://github.com/nats-io/nats.c/commit/4b64e7a3b41086b7442601d867815159b60bc6be).

Maybe that was a different time and trying to protect against early run of the dispatch code before the subscription was fully configured? Could it be that there was a chance that - since the proto is already sent - messages would be already coming and possibly user callback invoked even before the `natsConnection_Subscribe()` call returned?
So either let's leave it, or make sure that it is safe to remove.

levb replied:
Sorry I misunderstood your original comment. 1/5 Let's just leave it here as is, it's not expensive and been there for some time. I wonder if there needs to be an equivalent barrier (and what it would be) in the pool version, if the condition you described can happen there.

kozlovic commented on src/js.c:
```diff
@@ -1218,14 +1219,26 @@ _lookupStreamBySubject(const char **stream, natsConnection *nc, const char *subj
 
-void
-jsSub_free(jsSub *jsi)
+static void _destroyFetch(jsFetch *fetch)
```
formatting

kozlovic commented on src/js.c:
```diff
@@ -1218,14 +1219,26 @@ _lookupStreamBySubject(const char **stream, natsConnection *nc, const char *subj
+    NATS_FREE(fetch);
+}
+
+void jsSub_free(jsSub *jsi)
```
formatting

kozlovic commented on src/js.c:
```diff
@@ -2827,6 +2903,219 @@ js_PullSubscribe(natsSubscription **sub, jsCtx *js, const char *subject, const c
+    return NATS_UPDATE_ERR_STACK(s);
+}
+
+static bool _autoNextFetchRequest(jsFetchRequest *req, natsSubscription *sub, void *closure)
```
foramatting

kozlovic commented on src/js.c:
```diff
@@ -2827,6 +2903,219 @@ js_PullSubscribe(natsSubscription **sub, jsCtx *js, const char *subject, const c
+    natsSubscription *sub = NULL;
+    jsSub *jsi = NULL;
+    jsFetch *fetch = NULL;
+    jsFetchRequest defaultLifetime = {
```
Again, this init method may cause warnings on some OS'es.

levb replied:
0/5 this is a basic struct literal, since all fields are constants. I'd leave it as is for its better typography, fixed the other one.

kozlovic commented on src/js.c:
```diff
@@ -2827,6 +2903,219 @@ js_PullSubscribe(natsSubscription **sub, jsCtx *js, const char *subject, const c
+    // will not be active yet.
+    s = _subscribe(&sub, js, subject, durable, msgCB, msgCBClosure, true, jsOpts, opts, errCode);
+
+    IFOK(s, (fetch = NATS_CALLOC(1, sizeof(jsFetch))) == NULL ? NATS_NO_MEMORY : NATS_OK);
```
Inconvenient of that is that you are not calling `nats_setDefaultError(NATS_NO_MEMORY)`. So even though it is more code, you should have:
```
if (s == NATS_OK)
{
    fetch = NATS_CALLOC(1, sizeof(jsFetch));
    if (fetch == NULL)
        s = nats_setDefaultError(NATS_NO_MEMORY);
}
```

kozlovic commented on src/status.h:
```diff
@@ -130,6 +130,9 @@ typedef enum
 
     NATS_MISSED_HEARTBEAT,              ///< For JetStream subscriptions, it means that the library detected that server heartbeats have been missed.
 
+    NATS_LIMIT_REACHED,                 ///< Attempt to receive messages than allowed by the byte limit, for
```
Note that adding new values here requires that you bump the `NATS_VERSION_REQUIRED_NUMBER` in the main `CMakeLists.txt`. See `nats_CheckCompatibilityImpl()` for more information.

levb replied:
@kozlovic We are already set for 3.9 (next) in `main`, https://github.com/nats-io/nats.c/blob/4f19da67f801c61560c652bb84680ba021e042d7/CMakeLists.txt#L254-L259

Does it need to change?

kozlovic replied:
No, if it is already done.

kozlovic commented on test/test.c:
```diff
@@ -28636,6 +28639,495 @@ void test_JetStreamSubscribePull(void)
+    nats_clearLastError();
+
+    int noMessageTimeout = 40;
+    int messageArrivesImmediatelyTimeout = 20;
```
On my Windows VM, this test flaps a lot, and I wonder if you should not bump (either for all cases or just under windows) those 2 values,  maybe double: 80/40?

levb replied:
Updated, let's see?

kozlovic replied:
Even on macOS, with repeat of tests, I get:
```
== JetStreamSubscribePullAsync ==
#01 Start JS Server: PASSED
#02 Connect: PASSED
#03 Get context: PASSED
#04 Create stream: PASSED
#05 Create pull sub async (invalid args): PASSED
#06 AckNone ok: PASSED
#07 AckAll ok: PASSED
#08 Create push consumer: PASSED
#09 Try create pull sub from push consumer: PASSED
#10 Create pull bound failure: PASSED
#11 js_PullAsyncSubscribe for real, manual ack: PASSED
#12 Can't call NextMsg: PASSED
#13 Can't call Fetch: PASSED
#14 No messages yet, timeout in 80ms: PASSED
#15 Send a message: PASSED
#16 Arrives in under 40ms: PASSED
#17 No more messages yet, timeout in 80ms: PASSED
#18 Wait another so.Config.AckWait+100 = 400ms and see that the message is re-delivered, ACK it: PASSED
#19 Wait another so.Config.AckWait+100 = 400ms and see that the message is not re-delivered: PASSED
#20 Receive msg with header and no data comes through: PASSED
#21 single fetch NOWAIT, partially fulfilled NATS_TIMEOUT: PASSED
#22 multi-fetch NOWAIT, partially fulfilled NATS_TIMEOUT: PASSED
#23 fetch NOWAIT can get a NATS_NOT_FOUND immediately: PASSED
#24 fetch NOWAIT can get a NATS_NOT_FOUND on Nth request: PASSED
#25 single fetch NOWAIT, fulfilled msgs NATS_MAX_DELIVERED_MSGS: PASSED
#26 multi-fetch NOWAIT, fulfilled msgs NATS_MAX_DELIVERED_MSGS: PASSED
#27 single fetch, fulfilled msgs NATS_MAX_DELIVERED_MSGS: PASSED
#28 multi-fetch WAIT, fulfilled msgs NATS_MAX_DELIVERED_MSGS: PASSED
#29 MaxBytes: PASSED
#30 Fetch with expiration is fulfilled NATS_MAX_DELIVERED_MSGS: PASSED
#31 Fetch with a short expiration is partially fulfilled NATS_TIMEOUT: FAILED: status: 22, expected: 26
FAILED
*** TEST FAILED ***
```
or on Windows after the first run:
```
== JetStreamSubscribePullAsync ==
#01 Start JS Server: PASSED
#02 Connect: PASSED
#03 Get context: PASSED
#04 Create stream: PASSED
#05 Create pull sub async (invalid args): PASSED
#06 AckNone ok: PASSED
#07 AckAll ok: PASSED
#08 Create push consumer: PASSED
#09 Try create pull sub from push consumer: PASSED
#10 Create pull bound failure: PASSED
#11 js_PullAsyncSubscribe for real, manual ack: PASSED
#12 Can't call NextMsg: PASSED
#13 Can't call Fetch: PASSED
#14 No messages yet, timeout in 80ms: PASSED
#15 Send a message: PASSED
#16 Arrives in under 40ms: PASSED
#17 No more messages yet, timeout in 80ms: PASSED
#18 Wait another so.Config.AckWait+100 = 400ms and see that the message is re-delivered, ACK it: PASSED
#19 Wait another so.Config.AckWait+100 = 400ms and see that the message is not re-delivered: PASSED
#20 Receive msg with header and no data comes through: PASSED
#21 single fetch NOWAIT, partially fulfilled NATS_TIMEOUT: PASSED
#22 multi-fetch NOWAIT, partially fulfilled NATS_TIMEOUT: PASSED
#23 fetch NOWAIT can get a NATS_NOT_FOUND immediately: PASSED
#24 fetch NOWAIT can get a NATS_NOT_FOUND on Nth request: PASSED
#25 single fetch NOWAIT, fulfilled msgs NATS_MAX_DELIVERED_MSGS: PASSED
#26 multi-fetch NOWAIT, fulfilled msgs NATS_MAX_DELIVERED_MSGS: FAILED: timed out: 26
FAILED: onComplete has not been called
FAILED: status: 0, expected: 22
FAILED: msgs: 75, expected: 100
FAILED: subscription is still valid
FAILED
*** TEST FAILED ***
```

levb replied:
Resolved in favor of thread https://github.com/nats-io/nats.c/pull/785#discussion_r1725839401

kozlovic commented on test/test.c:
```diff
@@ -28636,6 +28639,495 @@ void test_JetStreamSubscribePull(void)
+            .expectedN = 2,
+        },
+        {
+            .name = "multi-fetch NOWAIT, fulfilled msgs NATS_MAX_DELIVERED_MSGS",
```
Again, on my VM, this test seem to fail almost all the time. Need to investigate more.

levb replied:
Resolved in favor of thread https://github.com/nats-io/nats.c/pull/785#discussion_r1725839401

kozlovic commented on test/test.c:
```diff
@@ -28636,6 +28639,495 @@ void test_JetStreamSubscribePull(void)
+    natsMutex_Unlock(args.m);
+
+    s = js_PullSubscribeAsync(&sub, js, "foo", "dur", _recvPullAsync, &args, &lifetime, &jsOpts, &so, &jerr);
+    testCond((s == NATS_OK) && _testBatchCompleted(&args, sub, 100, NATS_ERR, 0, false));
```
Again, tests often with s==NATS_OK, but function returns 0.

levb replied:
Sorry, did not understand this last one, Windows-specific?

kozlovic replied:
What I meant is that `s` is `NATS_OK`, what returns `0` is `_testBatchCompleted`.

levb replied:
Resolved in favor of thread https://github.com/nats-io/nats.c/pull/785#discussion_r1725839401

levb commented on src/js.c:
```diff
@@ -2827,6 +2903,220 @@ js_PullSubscribe(natsSubscription **sub, jsCtx *js, const char *subject, const c
+
+    *req = fetch->lifetime; // copy bytes, reading immutable data
+    req->Batch = want;
+    // FIXME discuss in PR - this seems wrong, we don't know how many bytes we will have
```
Note: still unresolved, and still not quite sure how

levb replied:
@kozlovic I am at a loss on how to properly resolve this, short of erroring on the side of over-asking and enforcing the limit in the client, which I do not like.

kozlovic replied:
I don't know either.

levb replied:
No longer allow `MaxBytes` and `KeepAhead` at the same time; With that, the math for the next request's MaxBytes should be accurate.

levb commented:
@kozlovic I resolved the comments where I think I added obvious fixes for the feedback, please resolve the remaining as you find fit.

I added tests for handling connect and reconnect while fetching, and made sure that OnFetchComplete handler is called if/when the sub is closed for whatever reason(s).

There are some flappers registering on Travis in `main` and here likely related to the refactoring. I am investigating, but that'll be a separate PR for it.

levb commented on examples/js-sub.c:
```diff
@@ -130,21 +175,50 @@ int main(int argc, char **argv)
+            jsOpts.PullSubscribeAsync.KeepAhead = 7;
+            jsOpts.PullSubscribeAsync.CompleteHandler = _completeFetchCb;
+
+            jsFetchRequest lifetime;
```
@kozlovic @Jarema @mtmk I chose to have a separate `lifetime` parameter as a shortcut for passing these options. 2/5 I am considering removing the extra parameter, and adding all these values to `jsOpts.PullSubscribeAsync`. (a) It would be more idiomatic; (b) arguably, CompleteHandler a more common use-case and they should be "level"-ed; and (c) the use of `jsFetchRequest` struct may be confusing since no such request will ever be made.

kozlovic replied:
You could. The thing I like about the options, is if we decide down the line that we need other option(s), then the function definition stays the same. We don't have to come up with a different function name, etc..
That being said, you could pass the parameters that seem essential at this time, and if in the future we need more, then add them through the options?

levb replied:
Replaced `lifetime` with more `jsOpts`, looks cleaner.

levb replied:
@Jarema @mtmk Can I please get your review on the API iteself (parameters, etc.) before we finalize this?

levb commented on test/test.c:
```diff
@@ -28741,6 +28744,792 @@ void test_JetStreamSubscribePull(void)
+    testCond(s == NATS_OK);
+
+    test("Check fetch completion, expect NATS_OK: ");
+    testCond(_testBatchCompleted(&args, sub, 500, NATS_OK, 1, false));
```
@kozlovic This is rather unfortunate: the present code invokes the callback with a `NATS_OK`, rather than `NATS_CONNECTION_DISCONNECTED` status. I would appreciate a suggestion on how to best pass that context into the callback [here](https://github.com/nats-io/nats.c/pull/785/files#diff-d04d35062495ef1ffc5dedde8c86073128c37a4e0594ee1a8cd3bee7e7552729R250) and [here](https://github.com/nats-io/nats.c/pull/785/files#diff-d04d35062495ef1ffc5dedde8c86073128c37a4e0594ee1a8cd3bee7e7552729R523)

kozlovic replied:
Sorry, those links don't work for me. So not sure what you are asking really.

levb commented on test/test.c:
```diff
@@ -28871,7 +28871,29 @@ _testBatchCompleted(struct threadArg *args, natsSubscription *sub, int waitMS, n
+    result = result && !natsSubscription_IsValid(sub);
+    if (!result)
+    {
+        if (s != NATS_OK)
```
@kozlovic Both failures should be fixed now.
- `Fetch with a short expiration is partially fulfilled NATS_TIMEOUT` was a genuine bug with the `ms`-`ns` confusion, compounded by a typo. It is now fixed (indirectly) with refactoring of `lifetime`->`jsOpts` which clarified the time units. Many thanks for catching it, it rarely failed elsewhere, and never(hundreds of retries!) on my Mac. The test itself didn't need to change much.
- `multi-fetch NOWAIT, fulfilled msgs NATS_MAX_DELIVERED_MSGS: FAILED: timed out: 26` I could not reproduce at all, but it appears that the wait timeout for the test was not long enough to fetch all of the messages, so I increased it.

Let me tidy up other remaining threads, I'll mark you for re-review when ready.

kozlovic commented on src/js.c:
```diff
@@ -3041,30 +3030,22 @@ js_PullSubscribeAsync(natsSubscription **newsub, jsCtx *js, const char *subject,
 
-        if (fetch->fetchSize == 0)
-            fetch->fetchSize = NATS_DEFAULT_ASYNC_FETCH_SIZE;
+#define _set(_f, _v, _nil, _def) fetch->_f = ((jsOpts != NULL) && (jsOpts->PullSubscribeAsync._v != _nil)) ? jsOpts->PullSubscribeAsync._v : _def
```
At this point, I wonder why not embed a PullSubscribeAsync option block inside the jsFetch and copy the user provided values (if jsOpts != NULL) into it?

levb replied:
ok.

kozlovic commented on src/nats.h:
```diff
@@ -1281,26 +1281,39 @@ typedef struct jsOptions
-                int KeepAhead;
+                // Lifetime of the subscription (completes when any one of the
+                // targets is reached).
+                int64_t                 TimeoutMillis;
```
nit: I know that it may sound obvious that `TimeoutMillis` means timeout in milliseconds, but I would have used `Timeout` but document above that the time is in milliseconds.

levb replied:
👍 I went a little overboard after confusing myself.

kozlovic commented on src/nats.h:
```diff
@@ -1281,26 +1281,39 @@ typedef struct jsOptions
+                void                    *CompleteHandlerClosure;
+
+                // Have server sends heartbeats to help detect communication failures.
+                int64_t                 HeartbeatMillis;
```
nit: Same here. For `Heartbeat`, if you feel that it is not telling enough, it could have been `HeartbeatInterval`.

kozlovic commented on test/test.c:
```diff
@@ -29286,16 +29288,17 @@ void test_JetStreamSubscribePullAsync_MissedHB(void)
     args.sum = 0;
+    natsMutex_Unlock(args.m);
+
+    // Let's make it wait for 20 seconds, and have HBs every 50ms
```
Still getting failures on Windows:
```
    Start 108: JetStreamSubscribePullAsync_MissedHB
    Test #108: JetStreamSubscribePullAsync_MissedHB ......***Failed  Required regular expression not found. Regex=[ALL PASSED
]  0.66 sec
Test using server executable: c:\test\nats-server.exe
Test using server executable: c:\test\nats-streaming-server.exe
Test server version: "nats-server: v2.10.18"

== JetStreamSubscribePullAsync_MissedHB ==
#01 Start JS Server: PASSED
#02 Connect: PASSED
#03 Get context: PASSED
#04 Create stream: PASSED
#05 Check invalid heartbeat : FAILED: status: 26, expected: 1
FAILED
*** TEST FAILED ***
```

kozlovic commented on test/test.c:
```diff
@@ -29354,15 +29357,15 @@ void test_JetStreamSubscribePullAsync_Unsubscribe(void)
+    s = js_PullSubscribeAsync(&sub, js, "foo", "dur", _recvPullAsync, &args, &jsOpts, NULL, &jerr);
     testCond(s == NATS_OK);
 
     test("Receive 1 message: ");
```
This test too:
```
    Test #110: JetStreamSubscribePullAsync_Unsubscribe ...***Failed  Required regular expression not found. Regex=[ALL PASSED
]  0.66 sec
Test using server executable: c:\test\nats-server.exe
Test using server executable: c:\test\nats-streaming-server.exe
Test server version: "nats-server: v2.10.18"

== JetStreamSubscribePullAsync_Unsubscribe ==
#01 Start JS Server: PASSED
#02 Connect: PASSED
#03 Get context: PASSED
#04 Create stream: PASSED
#05 Publish a few messages: PASSED
#06 Subscribe: PASSED
#07 Receive 1 message: FAILED
*** TEST FAILED ***
```

levb replied:
ok, will try to repro and read through the code/test, see if I can figure out what's causing these 2. If I can't reproduce anywhere else, I may need to ask you to run an instrumented version on your Windows.

levb replied:
@kozlovic Can you please try again? I increased a few more test wait timeouts, hopefully those were the cause of the failures on the slow VM...

kozlovic reviewed (changes requested):
(Going to review the last commit in another review)

kozlovic reviewed (commented):
Nothing to add to that one.

kozlovic commented on src/nats.h:
```diff
@@ -1206,6 +1206,79 @@ typedef struct jsFetchRequest
+ */
+typedef bool (*natsFetchNextHandler)(int *messages, int64_t *bytes, natsSubscription *sub, void *closure);
+
+typedef struct jsOptionsPullSubscribeAsync
```
You don't need to make it a separate typedef. Put it back the way it was in `jsOptions`:
```
struct jsOptionsPullSubscribeAsync
        {
                // Lifetime of the subscription (completes when any one of the
                // targets is reached).
                int64_t                 Timeout; // in milliseconds
...
} PullSubscribeAsync;
```
Then, in natsp.h:
```
typedef struct __jsFetch
{
    struct jsOptionsPullSubscribeAsync opts;
...
```
That will work.

levb replied:
Oh nice, I didn't realize you could re-use those nested struct definitions.

kozlovic commented on src/js.c:
```diff
@@ -2827,6 +2903,206 @@ js_PullSubscribe(natsSubscription **sub, jsCtx *js, const char *subject, const c
+
+    if (jsOpts != NULL)
+        fetch->opts = jsOpts->PullSubscribeAsync;
+    if (fetch->opts.FetchSize == 0)
```
In the previous code (where you used the _set() macro), was your intent to say that for non "0/NULL" values of the options (if options itself was not NULL), you would use them as-is, if not set some default values? Correct?

I believe this is what you are doing here, but wanted to make sure.

levb replied:
Clarified and removed the confusing defaulting here. Still need to ensure the default for FetchSize and Nextf

kozlovic reviewed (changes requested):
Small changes (no need for a typedef). The tests pass now on Windows.

levb commented:
@wallyqs FYI, many GHA jobs are still failing to download server images.

kozlovic reviewed (approved):
LGTM. By the way, we may want to add to the list of disabled Windows warnings in the CMakeLists.txt: `add_compile_options(/wd4204) # non-constant aggregate initializer`. This is for when we do something like: `myStruct s = {"a", i};` (when one of the initialization value is not a constant)

Jarema commented on examples/js-sub.c:
```diff
@@ -130,21 +171,48 @@ int main(int argc, char **argv)
-        if (pull)
+        if (pull && async)
+        {
+            jsOpts.PullSubscribeAsync.MaxMessages = (int) total;
```
This is different to what all other clients do.

As per https://github.com/nats-io/nats-architecture-and-design/blob/main/adr/ADR-37.md

We treat `no_wait == true` as a footgun that can have heavy impact on the server.

levb replied:
@Jarema please note that this PR/API should not be seen as the C version of JS simplification. It's main purpose is to fit in with the existing C APIs, to enable async usage pattern of the existing (`js_PullSubscribe`/`js_Fetch`/`js_FetchRequest`). The most immediate issue this PR solves is enabling the use of the thread pool for Pull subscriptions message delivery. Previously, the user had to start a thread of their own to accomplish this. So, this implementation attempts to stay idiomatic with the above (and other Subscribe) APIs.

Also, I did not attempt to de-couple the "Subscribe" vs [multiple] "Consume" operations because of how complex the subscription state logic already is in the C client. So, there is only one lifetime, that of the Fetch, and the sub is auto-unsubscribe if it's reached, or if a serious error happens(*see below).

As to "NoWait" - again, the idea was to enable the existing functionality (including `js_FetchRequest` which allows direct control of `NoWait`) in the async use case, using a thread pool. This was the rationale for keeping it visible.

levb replied:
Discussed offline.

Jarema commented on examples/js-sub.c:
```diff
@@ -130,21 +171,48 @@ int main(int argc, char **argv)
+            jsOpts.PullSubscribeAsync.MaxMessages = (int) total;
+            jsOpts.PullSubscribeAsync.NoWait = true;
+            jsOpts.PullSubscribeAsync.Timeout = 3600 * 1000; // 1 hour
+            jsOpts.PullSubscribeAsync.FetchSize = 17;
```
this is really low value. We use between 100 and 1k, depending on client and at what threshold it starts getting diminishing returns on throughput.

levb replied:
These are not the defaults, just example values. I'll change the example to explicitly set all values to their respective defaults, that'd be more intention-revealing. The default fetch size is 128. The rest are all 0s. 2/5 `KeepAhead` should default to 0, given non-trivial implications of setting it.

Jarema commented on examples/js-sub.c:
```diff
@@ -130,21 +171,48 @@ int main(int argc, char **argv)
+            jsOpts.PullSubscribeAsync.NoWait = true;
+            jsOpts.PullSubscribeAsync.Timeout = 3600 * 1000; // 1 hour
+            jsOpts.PullSubscribeAsync.FetchSize = 17;
+            jsOpts.PullSubscribeAsync.KeepAhead = 7;
```
We agreed that the default threshold for triggering new PR is 50% of initial batch.

https://github.com/nats-io/nats-architecture-and-design/blob/main/adr/ADR-37.md#defaults-and-constraints

levb replied:
Updated to use the actual defaults in the example, with comments on what these are.

Jarema commented on examples/js-sub.c:
```diff
@@ -130,21 +171,48 @@ int main(int argc, char **argv)
+            jsOpts.PullSubscribeAsync.CompleteHandler = _completeFetchCb;
+            jsOpts.PullSubscribeAsync.CompleteHandlerClosure = NULL;
+
+            // Uncomment to use a 1 second heartbeat.
```
Heartbeat should be half of expires.

levb replied:
I am not sure what this means, Expires is the lifetime of the sub.

Jarema commented on src/jsm.c:
```diff
@@ -1978,7 +1978,7 @@ js_directGetMsgToJSMsg(const char *stream, natsMsg *msg)
         && (natsMsgHeader_Get(msg, STATUS_HDR, &val) == NATS_OK))
     {
-        if (strcmp(val, NOT_FOUND_STATUS) == 0)
+        if (strcmp(val, HDR_STATUS_NOT_FOUND_404) == 0)
```
All the errors should be communicated to the user (including `missing heartbeats` which should happen after missing hearbeat set duration x2), but none should terminate the proccess.

levb replied:
Again this came out of preserving the current APIs and attempting at consistency with them (`js_Fetch` is effectively unchanged by this PR).

levb replied:
Discussed offline, no change

Jarema commented on src/nats.h:
```diff
@@ -1243,6 +1278,44 @@ typedef struct jsOptions
+                // If NoWait is set, the subscription will receive the messages
+                // already stored on the server subject to the limits, but will
+                // not wait for more messages.
+                bool                    NoWait;
```
this should be not exposed as a public API.
Rather have two different methods for fetching with and without wait.

The reason being - expires and no wait has different behaviors:
expires set + no_wait - will try to get all messages up to the batch, or the number currently available
no expires + no_wait - will always wait for first message. Even if there are currently any on the server.

levb replied:
As discussed offline, added an extra comment to clarify, but keeping the flag.

Jarema commented on examples/js-sub.c:
```diff
@@ -130,21 +171,49 @@ int main(int argc, char **argv)
+            // Defalut values, change as needed.
+            jsOpts.PullSubscribeAsync.FetchSize = 128;  // ask for 128 messages at a time
+            jsOpts.PullSubscribeAsync.NoWait = false;
+            jsOpts.PullSubscribeAsync.Timeout = 0;      // for the entire subscription, in milliseconds
```
This is fine,
But I think that each pull request should have 30s expires default value.
Otherwise if server looses the Pull Requst, client will never know about it.

Similarly, HeartBeats should be enabled by default, with a value of 15 seconds (half of expires, which is maximum).

levb replied:
@Jarema First, 1/5 I don't think that subscription-level Expires/Timeout should be defaulted on. By default, a sub should live until Unsubscribe, that's the idiom. A separate FetchTimeout (was Batch a better name?) would solve this issue but also add complexity and potential confusion.

Second, if the heartbeats are ON, and the server loses a pull request, the client _will_ know about because of the missing heartbeat error. So, if we make heartbeat ON by default at some number of seconds why add a default timeout?

Jarema replied:
I agree on not having timeout on subscription. In fact, none other clients have that option at all.
If you want to stop subscription, stop it explicitly, or do drain. We never do timeouts like thoge.

What I meant however, is `expires` value for each `pull requst`, and `heartbeats` for those pull requests too.
